src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -0,0 +1,338 @@
+
+                            var (cachedCommitHistory, lruOrder) = cachedValue;
+                            foreach (var cachedCommit in cachedCommitHistory)
+                            {
+                                result.Add(commitsBySha[cachedCommit]);
+                            }
+                            commitCache[(commit.Sha.a, blob)] = (cachedCommitHistory, _nextLruOrder--);
+                            break;
+                        }
+                    }
+                }
+
+                var singleParent = false;
+                var parentCount = Math.Min(MaxParentBlob, commit.Parents.Length);
+                var add = parentCount == 0 && blob != 0;
+
+                for (var i = 0; i < parentCount; i++)
+                {
+                    parentBlobs[i] = GetBlob(commit.Parents[i].Tree, pathSegments);
+                    if (parentBlobs[i] == blob)
+                    {
+                        // and it was TREESAME to one parent, follow only that parent.
+                        // (Even if there are several TREESAME parents, follow only one of them.)
+                        commitsToFollow.Add((commit.Parents[i], blob));
+                        singleParent = true;
+                        break;
+                    }
+                }
+
+                if (!singleParent)
+                {
+                    // Otherwise, follow all parents.
+                    for (var i = 0; i < parentCount; i++)
+                    {
+                        add = true;
+                        commitsToFollow.Add((commit.Parents[i], parentBlobs[i]));
+                    }
+                }
+
+                if (add)
+                {
+                    result.Add(commit);
+                }
+            }
+
+            if (updateCache)
+            {
+                lock (commitCache)
+                {
+                    _cacheUpdated = true;
+                    commitCache.Add((headCommit.Sha.a, headBlob), (result.Select(c => c.Sha.a).ToArray(), 0));
+                }
+            }
+
+            return result.Select(c => c.GitCommit).ToList();
+        }
+
+        public IReadOnlyList<KeyValuePair<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>> BuildCache()
+        {
+            if (_cacheUpdated)
+            {
+                // There is a race condition in Linq ToList() method, use ConcurrentDictionary.ToArray() to create a snapshot
+                // https://stackoverflow.com/questions/11692389/getting-argument-exception-in-concurrent-dictionary-when-sorting-and-displaying
+                return _commitCache.ToArray();
+            }
+
+            return null;
+        }
+
+        public void Dispose()
+        {
+            var repo = Interlocked.Exchange(ref _repo, IntPtr.Zero);
+            if (repo != IntPtr.Zero)
+            {
+                git_repository_free(_repo);
+            }
+
+            GC.SuppressFinalize(this);
```

Comment:
Seems like we should follow the pattern as below?
https://docs.microsoft.com/en-us/dotnet/api/system.gc.suppressfinalize?view=netframework-4.7.2
https://stackoverflow.com/questions/151051/when-should-i-use-gc-suppressfinalize

Reply:
we are using sealed class, no need follow the complex pattern